### PR TITLE
WGSL support

### PR DESCRIPTION
**Pull request by dparnell**
Hi,

  I spent a little bit of time today adding WGSL support via the naga crate.
So far it seems to be working as expected.

Daniel

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -243,60 +333,109 @@ fn compile(
...
+                                    Ok(feedback)
+                                }
+                                Err(e) => Err(format!("{:?}", e))
+                            }
+                        },
+                        Err(e) => Err(format!("{:?}", e))
+                    }
+                },
+                Err(e) => {
+                    e.emit_to_stderr(src);
```
Comment (PENGUINLIONG) [inline src/lib.rs]: Could you attach the source before the error messages emitted next line? The users usually want more control over the output streams.

File: src/lib.rs [hunk 2]
```diff
@@ -218,20 +286,42 @@ fn parse_compile_cfg(
...
+                #[cfg(feature = "shaderc")]
+                { cfg.vulkan_version = EnvVersion::Vulkan1_1; }
+                #[cfg(feature = "naga")]
+                { cfg.naga_spirv_version = (1, 1); }
+            },
+            "vulkan1_2" => {
+                #[cfg(feature = "shaderc")]
+                { cfg.vulkan_version = EnvVersion::Vulkan1_2; }
+                #[cfg(feature = "naga")]
+                { cfg.naga_spirv_version = (1, 2); }
```
Comment (PENGUINLIONG) [inline src/lib.rs]: Vulkan 1.2 uses SPIR-V 1.5.

File: src/lib.rs [hunk 3]
```diff
@@ -218,20 +286,42 @@ fn parse_compile_cfg(
...
+                #[cfg(feature = "shaderc")]
+                { cfg.vulkan_version = EnvVersion::Vulkan1_0; }
+                #[cfg(feature = "naga")]
+                { cfg.naga_spirv_version = (1, 0); }
+            },
+            "vulkan1_1" => {
+                #[cfg(feature = "shaderc")]
+                { cfg.vulkan_version = EnvVersion::Vulkan1_1; }
+                #[cfg(feature = "naga")]
+                { cfg.naga_spirv_version = (1, 1); }
```
Comment (PENGUINLIONG) [inline src/lib.rs]: Vulkan 1.1 uses SPIR-V 1.3.

File: src/lib.rs [hunk 4]
```diff
@@ -108,38 +108,88 @@
...
+
+#[cfg(not(any(feature = "shaderc", feature = "naga")))]
+compile_error!("at least one compiler feature must be specified");
+
+enum InputSourceLanguage {
+    #[cfg(feature = "shaderc")]
+    ShaderC(SourceLanguage),
+
+    #[cfg(feature = "naga")]
+    WGSL
```
Comment (PENGUINLIONG) [inline src/lib.rs:128]: Sounds like a mix of compilation backends and source languages. It's probably better to flatten `InputSourceLanguage::ShaderC(HLSL)` and `InputSourceLanguage::ShaderC(GLSL)` to `InputSourceLanguage::HLSL` and `InputSourceLanguage::GLSL`

**Review by PENGUINLIONG — CHANGES_REQUESTED** [review]
Thanks for your contribution, that's a lot of work! Except for several minor issues I think its good to go. :)

File: src/lib.rs (hunk 1 above)
Comment (dparnell) [reply src/lib.rs]: this should be fixed in the most recent push

File: src/lib.rs (hunk 2 above)
Comment (dparnell) [reply src/lib.rs]: updated

File: src/lib.rs (hunk 3 above)
Comment (dparnell) [reply src/lib.rs]: and this one too

File: src/lib.rs (hunk 4 above)
Comment (dparnell) [reply src/lib.rs:128]: and done

**Comment by dparnell** [comment]
the most recent push should resolve the requested changes

**Comment by PENGUINLIONG** [comment]
Thanks for your contribution, that's a lot of work! A new version should be ready to go when I get back home. :)

**Comment by dparnell** [comment]
It was my pleasure :)

**Pull request merged**